### ISSUE TITLE
Invalid syntax generated in assignment of command substitutions

### ISSUE THREAD

**Issue by Xophmeister** [tweag/topiary#201]
**Describe the bug**
In the following contexts, an errant soft spaced line is inserted after the `=` sign in variable assignment:
* String of...
* Command substitution containing...
* Whitespace

:warning: This renders syntactically incorrect Bash.

**To Reproduce**
Run the following through the Bash formatting rules:

Single-line
```bash
a="$(foo bar)"
```

Mulit-line
```bash
a="$(
  foo
)"
```

**Expected behavior**
The above code example should not be changed by the formatter. Currently, it is formatted as follows:

Single-line
```bash
a= "$(foo bar)"
```

Multi-line
```bash
a=
"$(
  foo
)"
```

**Environment**
 - OS name + version: NixOS 22.05
 - Version of the code: 12f1894

**Additional context**
Not for v0.1

**Comment by aspiwack** [comment]
Should this be solved before v0.1?

**Comment by Xophmeister** [comment]
> Should this be solved before v0.1?

Ideally, yes, absolutely. It's not clear what's causing this behaviour, but I agree that anything that renders syntactically incorrect output ought to be corrected as a matter of priority.

That said, I wouldn't say it's a show-stopper as there is a viable workaround. (Although, quite legitimately, formatter quirks should _not_ force you to write code in a specific way!) While it's normally highly advised to quote variable assignment, it isn't necessary for command substitutions (e.g., Shellcheck doesn't mark it as a lint error). Forgoing the quoting is enough to relax the constraints that manifest this bug.

**Comment by aspiwack** [comment]
Let's put it in v0.1 for now. But it it proves too long to fix, we'll document the limitation instead.

**Comment by Xophmeister** [comment]
The errant space in the single-line context is coming from this rule:
https://github.com/tweag/topiary/blob/12f1894da2fdd16d4ba2531c942dbf6fd4aaed90/languages/bash.scm#L347-L354

The errant line feed in the multi-line context is coming from these rules:
https://github.com/tweag/topiary/blob/12f1894da2fdd16d4ba2531c942dbf6fd4aaed90/languages/bash.scm#L363-L372

**Comment by Xophmeister** [comment]
Addendum: The single-line space is coming from many rules that prepend a space (arguments, lists, pipelines and redirections).

**Comment by Xophmeister** [comment]
The general problem, I believe, is that `(string)`s are leaf nodes, so any formatting directives that happen inside them are "pushed out" to their edges.

**Comment by aspiwack** [comment]
Could this be fixed by a clever application of a `@do_nothing` tag?

**Comment by Xophmeister** [comment]
> Could this be fixed by a clever application of a `@do_nothing` tag?

I don't think so. You can target strings that contain named children easily enough with `(string (_))`, but there's no way to target a pure literal. Something like the following doesn't do the job:

```scm
(string
  (_) @do_nothing
) @leaf
```

I can fix (and have, in 431aa10) the single-line context by appending spaces, rather than prepending. A space will get added after the string, I believe, but this will be cleaned up in post-processing.

The multi-line context is proving to be a little more trying...
